cirq-google/cirq_google/engine/stream_manager.py

```diff
@@ -75,5 +96,276 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
+                )
+                async for response in response_iterable:
+                    self._response_demux.publish(response)
+            except asyncio.CancelledError:
+                # Handled in _manage_stream_cancel() callback.
+                # Handling here does not work because the ResponseDemux may have been reset in the
+                # duet thread.
+                break
+            except BaseException as e:
+                # TODO(#5996) Close the request iterator to close the existing stream.
+                # Note: the message ID counter is not reset upon a new stream.
+                self._response_demux.publish_exception(e)  # Raise to all request tasks
+
+    def _manage_stream_cancel(self):
+        def cancel(future) -> None:
+            if future.cancelled():
+                self._response_demux.publish_exception(asyncio.CancelledError())
+
+        return cancel
+
+    async def _manage_execution(
+        self, project_name: str, program: quantum.QuantumProgram, job: quantum.QuantumJob
+    ) -> Union[quantum.QuantumResult, quantum.QuantumJob]:
+        """The execution coroutine, an asyncio coroutine to manage the lifecycle of a job execution.
+
+        This coroutine sends QuantumRunStream requests to the request iterator and receives
+        responses from the ResponseDemux.
+
+        It initially sends a CreateQuantumProgramAndJobRequest, and retries if there is a retryable
+        error by sending another request. The exact request type depends on the error.
+
+        There is one execution coroutine per running job submission.
+        """
+        # Construct requests ahead of time to be reused for retries.
+        create_program_and_job_request = quantum.QuantumRunStreamRequest(
+            parent=project_name,
+            create_quantum_program_and_job=quantum.CreateQuantumProgramAndJobRequest(
+                parent=project_name, quantum_program=program, quantum_job=job
+            ),
+        )
+
+        while self._request_queue is None:
+            # Wait for the stream coroutine to start.
+            # Ignoring coverage since this is rarely triggered.
+            await asyncio.sleep(1)  # coverage: ignore
+
+        current_request = create_program_and_job_request
+        while True:
+            try:
+                current_request.message_id = self._generate_message_id()
+                response_future = self._response_demux.subscribe(current_request.message_id)
+                await self._request_queue.put(current_request)
+                response = await response_future
+
+            # Broken stream
+            except google_exceptions.GoogleAPICallError as e:
+                if _is_retryable_error(e):
+                    # Retry
+                    current_request = _to_get_result_request(create_program_and_job_request)
+                    continue
+                    # TODO(#5996) add exponential backoff
+                raise e
```

Comment:
nit: I'd prefer to handle the error case first, e.g.
```suggestion
                if not _is_retryable_error(e):
                    raise e
                    
                # Retry
                current_request = _to_get_result_request(create_program_and_job_request)
                continue
                # TODO(#5996) add exponential backoff
```
Could also consider putting the success case below in an `else` clause of the `try-except`, which would avoid the need for the `continue`.